### PR TITLE
Memory: Object dispose

### PR DESCRIPTION
**Pull request by tomorrowevening**
**Description**

A thorough way of disposing objects for memory.
Additional events have been added for when Scenes and the WebGLRenderer have been disposed.

### CODE REVIEW

File: src/materials/Material.js [hunk 1]
```diff
@@ -506,6 +506,18 @@ class Material extends EventDispatcher {
 
 	dispose() {
 
+		for ( const i in this ) {
+
+			const value = this[ i ];
+
+			if ( value !== null ) {
+
+				if ( value.isTexture ) value.dispose();
```
Comment (Mugen87) [inline src/materials/Material.js:515]: You are making assumptions in this PR which are not correct. E.g. textures can be shared in materials, shader so it is no correct to call `dispose()` on them if the material gets disposed of.

File: src/objects/Mesh.js [hunk 2]
```diff
@@ -72,6 +72,16 @@ class Mesh extends Object3D {
 
 	}
 
+	dispose() {
+
+		this.geometry.dispose();
+
+		this.material.dispose();
```
Comment (Mugen87) [inline src/objects/Mesh.js:79]: Similar issue with textures. Materials and geometries can be shared with other meshes so it is not correct to call `dispose()` on these object here.

File: src/renderers/WebGLRenderer.js [hunk 3]
```diff
@@ -645,6 +645,12 @@ class WebGLRenderer {
 
 			animation.stop();
 
+			if ( typeof __THREE_DEVTOOLS__ !== 'undefined' ) {
+
+				__THREE_DEVTOOLS__.dispatchEvent( new CustomEvent( 'dispose', { detail: this } ) );
```
Comment (Mugen87) [inline src/renderers/WebGLRenderer.js:650]: AFAIK, are not further developed anymore so they won't use the `dispose()` event.

The renderer should only emit this event if it is actually consumed by the three dev tools.

File: src/core/Object3D.js [hunk 4]
```diff
@@ -1015,6 +1015,19 @@ class Object3D extends EventDispatcher {
 
 	}
 
+	dispose() {
+
+		while ( this.children.length > 0 ) {
+
+			const child = this.children[ 0 ];
+			child.dispose();
```
Comment (Mugen87) [inline src/core/Object3D.js:1023]: Same with textures. Only because a 3D object is removed, it does not _always_ mean its children are removed as well.

**Comment by Mugen87** [comment]
Please do not updated build files. There should be no need to update screenshots. The one that you have updated are almost all black which isn't right.

File: src/renderers/WebGLRenderer.js (hunk 3 above)
Comment (tomorrowevening) [reply src/renderers/WebGLRenderer.js:650]: Yea I figured, however I know [Three Tools](https://three.tools/) uses this and it was mostly for consistency, similar to `Texture`. I mostly wanted the `Scene` to dispatch when it gets disposed  https://github.com/mrdoob/three.js/blob/ac95b71239d45ae9fd489d6e13b2729803d97143/src/textures/Texture.js#L209

File: src/core/Object3D.js (hunk 4 above)
Comment (tomorrowevening) [reply src/core/Object3D.js:1023]: The point is to clear all Object's memory, like a `Scene`. If assets are shared, why not set to null before calling `dispose`?

**Comment by tomorrowevening** [comment]
@Mugen87 Thanks! Yea I dunno why Github was throwing errors regarding the screenshots & CI/E2E, but I'll keep that in mind for future PRs. Is there anything from this PR that has merit or should I close?

**Comment by kristiker** [comment]
I guess the broken screenshots indicate the breakage brought by code changes.

File: src/core/Object3D.js (hunk 4 above)
Comment (Mugen87) [reply src/core/Object3D.js:1023]: There is no point in calling `dispose()` on `Object3D` level if you can't call `dispose()` on materials and geometries.

**Comment by Mugen87** [comment]
@tomorrowevening If you want this kind of `dispose()` behavior in your app, I suggest you implement a helper in app level. The suggested approach is however inappropriate as a default behavior.

**Pull request closed without merging**